packages/ckeditor5-ui/src/textarea/textareaview.ts

```
@@ -7,39 +7,189 @@
+		const fullTextValueClone = getTextareaElementClone( viewElement, viewElement.value );
+		const singleLineContentStyles = singleLineContentClone.ownerDocument.defaultView!.getComputedStyle( singleLineContentClone );
+
+		const verticalPaddings = parseFloat( singleLineContentStyles.paddingTop ) + parseFloat( singleLineContentStyles.paddingBottom );
+		const borders = getBorderWidths( singleLineContentClone );
+		const lineHeight = parseFloat( singleLineContentStyles.lineHeight );
+		const verticalBorder = borders.top + borders.bottom;
+
+		const singleLineAreaDefaultHeight = new Rect( singleLineContentClone ).height;
+		const numberOfLines = Math.round( ( fullTextValueClone.scrollHeight - verticalPaddings ) / lineHeight );
+
+		const maxHeight = this.maxRows * lineHeight + verticalPaddings + verticalBorder;
+
+		// There's a --ck-ui-component-min-height CSS custom property that enforces min height of the component.
+		// This min-height is relevant only when there's one line of text. Other than that, we can rely on line-height.
+		const minHeight = numberOfLines === 1 ? singleLineAreaDefaultHeight : this.minRows * lineHeight + verticalPaddings + verticalBorder;
+
+		// The size of textarea is controlled by height style instead of rows attribute because event though it is
+		// a more complex solution, it is immune to the layout textarea has been rendered in (gird, flex).
+		this._height = Math.min(
+			Math.max(
+				Math.max( numberOfLines, this.minRows ) * lineHeight + verticalPaddings + verticalBorder,
+				minHeight
+			),
+			maxHeight
+		);
+
+		if ( shouldScroll ) {
+			viewElement.scrollTop = viewElement.scrollHeight;
+		}
+
+		singleLineContentClone.remove();
+		fullTextValueClone.remove();
+	}
+
+	/**
+	 * Validates the {@link #minRows} and {@link #maxRows} properties and warns in the console if the configuration is incorrect.
+	 */
+	private _validateMinMaxRows() {
+		if ( this.minRows > this.maxRows ) {
```

I'm wondering if we should throw an error. WDYT?